File: solutions/ruby/resistor-color-trio/1/resistor_color_trio.rb
class ResistorColorTrio

  TEMPLATE = 'Resistor value: %<value>i %<prefix>sohms'

  BAND = {
     'black' => 0,
     'brown' => 1,
       'red' => 2,
    'orange' => 3,
    'yellow' => 4,
     'green' => 5,
      'blue' => 6,
    'violet' => 7,
      'grey' => 8,
     'white' => 9
  }

  private

  attr_writer :colors

  def initialize(colors)
    self.colors = colors.take(3)
    base_value = (BAND[self.colors[0]] * 10) + BAND[self.colors[1]]
    multiplier = 10**BAND[self.colors[2]]
    @total_ohms = base_value * multiplier
  end

  public

  attr_reader :colors

  def label
    value = @total_ohms
    prefix = ''

    if value >= 1_000_000_000
      value /= 1_000_000_000
      prefix = 'giga'
    elsif value >= 1_000_000
      value /= 1_000_000
      prefix = 'mega'
    elsif value >= 1_000
      value /= 1_000
      prefix = 'kilo'
    end

    TEMPLATE % { value: value, prefix: prefix }
  end

end

File: solutions/typescript/resistor-color-trio/1/resistor-color-trio.ts
const PREFIXES = ['', 'kilo', 'mega', 'giga'];

const COLORS = [
    'black',
    'brown',
    'red',
    'orange',
    'yellow',
    'green',
    'blue',
    'violet',
    'grey',
    'white',
] as const;

export type Color = (typeof COLORS)[number];

export function decodedResistorValue(
    colors: [Color, Color, Color, ...Color[]],
): string {
    const [tens, ones, multiplier] = colors
        .slice(0, 3)
        .map(color => COLORS.indexOf(color));

    let base = tens * 10 + ones;
    let zeros = multiplier;

    if (base > 0 && base % 10 == 0) {
        base /= 10;
        zeros += 1;
    }

    let prefix_index = (zeros / 3) | 0;
    let remaining_zeros = zeros % 3;

    let final_value = base * 10 ** remaining_zeros;
    let prefix = PREFIXES[prefix_index];

    return `${final_value} ${prefix}ohms`;
}
